### DISCUSSION TITLE
[v3] #[Reactive] properties & Eloquent lazy loading

### DISCUSSION THREAD

**Discussion by mcpuishor** [livewire/livewire#6090] [2. Show and tell]
I'm working on this component to display an order with its order lines and notes posted by users. The basic structure looks like this: 

```
Order -> hasMany -> Item as (items)
Item -> hasMany -> Note as (notes)
Note -> belongsTo -> User (as posteBy)
```

Items of an order are loaded using a query to Eloquent model:

```
$this->lines = Item::query()
           ->where('order_id', $this->order->id)
           ->with(['notes' => fn($q) => $q->latest()])
           ->get();
```

When displaying in a Livewire v3 component I created a component for displaying the order items. 

```
@forelse($lines as $orderline)
 <livewire:orders.order-line :$orderline wire:key="{{ $orderline->id }}" />
@empty
 This order has no lines. Please contact admin and check with your stock controller.
@endforelse
```

The `Orders\OrderLine` component has a property defined like this: 

```
class OrderLine extends Component
{
    use LivewireAlert;

    #[Reactive] public Item $orderline;
}
```

with a Blade file that takes the collection of notes for each item and loops through them. 

```
@foreach ( $orderline->notes as $note)
        {{ $note->message }}
        Posted by {{ $note->postedBy->name }}
@endforeach
```

For 3 days I banged my head against all possible walls trying to understand why when displaying the notes the page would crash with the following exception:

```
Cannot mutate reactive prop [orderline] in component: [orders.order-line]
```

I tried all possible things to remove, to refactor. There isn't much help online as we are still in the beta phase of Livewire 3... So banging against the walls I did, for three days now. 

Finally, after removing one attribute at the time to understand where anything is actually being mutated, I figure it out. 

### Reactive properties don't play well with Eloquent Lazy load

I was eager loading most of the relationships with the items in the order, but one: `note.postedBy` . I simply assumed that lazy load would work and the relationship would be loaded when needed, as it would do in a normal Blade file. Optimizing the N+1 queries was not the priority at the moment.

Therefore, when the Livewire child component `orders\order-line` would be displayed, it mutates the reactive property `orderline` by lazy loading the `notes.postedBy` relationship. 

Once realized that, I've eager loaded the `notes.postedBy` relationship when loading the items and all was fixed.

**Comment by joshhanley** [comment]
@mcpuishor thanks for sharing!

**Comment by robsontenorio** [comment]
@mcpuishor  Even with eager loading I am having this issue. I am on latest `3.0.5` as described here https://github.com/livewire/livewire/discussions/6813

It was suggested to use a custom Synthesizer. But , it would be too much boilerplate for a simple use case. And I see you are using collections.

Dou have any workaround?

**Reply by mcpuishor** [reply]
I had a quick look at the code, IMHO you don't need a nested component to begin with. Simply render the collection in your main component (if need be, with the help of a Blade component).


You get the error  because the collection of users gets mutated every time you refresh the list (when performing a search). 

If you insist on having nested components (probably for additional functionality), I suggest you make every user record a nested Livewire component.

**Reply by robsontenorio** [reply]
@mcpuishor  I just over simplified the example. The nested list component covers the scenario where the list would have a delete button on each row. And delete logic belongs to that nested list component. And it is more obvious to handle that logic within a livewire component the encapsulates everything.

Semantics asks for a "list component" instead a "component per row", but unfortunately it seems the only option.

**Reply by mcpuishor** [reply]
Sometimes the Occam's razor might be useful: if  you have two competing ideas to explain the same phenomenon, you should prefer the simpler one. 

Even a scenario where list elements have additional actions, can be handled by the main component. Just bind to a method with a parameter.

Because altering the collection inside the nested component will trigger a "mutability" (is that a word?) exception in the parent.

This was my main takeaway from this: reactive properties should not be altered in any way by the child component, but only by the parent, inside the parent.

**Reply by mcpuishor** [reply]
A similar situation is discussed and solved in #6033
